### PR TITLE
Checkjs implies allowjs

### PR DESCRIPTION
**Pull request by sandersn**
If it's not explicitly provided.

Note that it's still an error to explicitly provide `checkJs: true, allowJs: false`.

This change makes the compiler options easier to use and to describe.

### CODE REVIEW

**Comment by sandersn** [comment]
@DanielRosenwasser @RyanCavanaugh @orta Do you think the new behaviour is reasonable?

@sheetalkamat @weswigham Is `createProgram` the right place to put this?

**Comment by weswigham** [comment]
For other defaulted flags dependent on other flags, we have something like a `getStrictCompilerOptionValue` helper that we use everywhere rather than checking the compiler options directly (no, I don't think `createProgram` handles all inputs - `transpileModule` is notable for probably being unaffected). You should probably add a `getAllowJSCompilerOption` helper that we use everywhere rather than checking the flags directly.

**Comment by sandersn** [comment]
I thought that might be the case. I switched to an accessor function `getAllowJSCompilerOption` instead.

File: src/compiler/program.ts [hunk 1]
```diff
@@ -3160,7 +3160,7 @@ namespace ts {
                 createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_when_option_target_is_ES3, "useDefineForClassFields");
             }
 
-            if (options.checkJs && !options.allowJs) {
+            if (options.checkJs && !getAllowJSCompilerOption(options)) {
```
Comment (sheetalkamat) [inline src/compiler/program.ts:3163]: This seems incorrect condition. You want to check if options.allowJs === false here instead

**Comment by DanielRosenwasser** [comment]
I'm not so sure how I feel about this. But I can't think of any reason why it's bad.

File: src/compiler/program.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/program.ts:3163]: Duh, thanks for the catch.

Comment (sandersn) [reply src/compiler/program.ts:3163]: No, wait, that undoes the loosening from this PR. THe current code only errors when `checkJs: true` but `allowJs: false` explicitly.

**Comment by sandersn** [comment]
Chatted with @RyanCavanaugh, he couldn't either.

He also pointed out that maybe allowJs should default to true now. I'll investigate that next.

File: src/compiler/program.ts (hunk 1 above)
Comment (ExE-Boss) [reply src/compiler/program.ts:3163]:
```suggestion
            if (options.checkJs && options.allowJs === false) {
```
will also work and won’t require a function call.

**Pull request merged**